### PR TITLE
HackStudio: Fix bugs around the unsaved files warning

### PR DESCRIPTION
**Pull request by sk-spr**
This PR is intended to fix two related bugs that occurred when using the dialog that warns of unsaved files. The first issue is that when this dialog was spawned while NewProjectDialog was open, the two windows would fight and neither would be clickable. 
The second issue was that when saving all files through said dialog, any files that had never been saved and thus had no filename would crash HackStudio. This was fixed by adding a "Save as" dialog to the save() function of EditorWrapper.

### CODE REVIEW

File: Userland/DevTools/HackStudio/Dialogs/NewProjectDialog.cpp [hunk 1]
```diff
@@ -93,6 +94,7 @@ NewProjectDialog::NewProjectDialog(GUI::Window* parent)
         if (path.has_value())
             m_create_in_input->set_text(path.value().view());
     };
+    verify_cast<HackStudioWidget>(parent->main_widget())->m_new_project_dialog = RefPtr<NewProjectDialog>(this);
```
Comment (ADKaster) [inline Userland/DevTools/HackStudio/Dialogs/NewProjectDialog.cpp:97]: This is not a good way to attach this data to the main widget.

In HackStudioWidget.cpp, we have this:

```
NonnullRefPtr<GUI::Action> HackStudioWidget::create_new_project_action()
{
    return GUI::Action::create("&Project...", Gfx::Bitmap::try_load_from_file("/res/icons/16x16/hackstudio-project.png"sv).release_value_but_fixme_should_propagate_errors(), [this](const GUI::Action&) {
        auto dialog = NewProjectDialog::construct(window());
        dialog->set_icon(window()->icon());
        auto result = dialog->exec();

        if (result == GUI::Dialog::ExecResult::OK && dialog->created_project_path().has_value())
            open_project(dialog->created_project_path().value());
    });
}
```
Just assign `this->m_new_project_dialog = dialog;` in this function, if it's actually correct to save it.

File: Userland/DevTools/HackStudio/Dialogs/NewProjectDialog.h [hunk 2]
```diff
@@ -21,14 +21,15 @@ namespace HackStudio {
 class NewProjectDialog : public GUI::Dialog {
     C_OBJECT(NewProjectDialog);
 
+    virtual ~NewProjectDialog() override = default;
```
Comment (ADKaster) [inline Userland/DevTools/HackStudio/Dialogs/NewProjectDialog.h:24]: This change is not needed, please revert this file

File: Userland/DevTools/HackStudio/HackStudioWidget.cpp [hunk 3]
```diff
@@ -1610,8 +1610,8 @@ HackStudioWidget::ContinueDecision HackStudioWidget::warn_unsaved_changes(String
 {
     if (!any_document_is_dirty())
         return ContinueDecision::Yes;
-
-    auto result = GUI::MessageBox::show(window(), prompt, "Unsaved changes"sv, GUI::MessageBox::Type::Warning, GUI::MessageBox::InputType::YesNoCancel);
+    auto dialog = verify_cast<HackStudioWidget>(window()->main_widget())->m_new_project_dialog;
```
Comment (ADKaster) [inline Userland/DevTools/HackStudio/HackStudioWidget.cpp:1613]: we are in HackStudioWidget here, you can just access this via `m_new_project_dialog` directly

File: Userland/DevTools/HackStudio/HackStudioWidget.h [hunk 4]
```diff
@@ -81,6 +82,7 @@ class HackStudioWidget : public GUI::Widget {
 
     void update_current_editor_title();
     void update_window_title();
+    RefPtr<NewProjectDialog> m_new_project_dialog;
```
Comment (ADKaster) [inline Userland/DevTools/HackStudio/HackStudioWidget.h:85]: this should be private

File: Userland/DevTools/HackStudio/Dialogs/NewProjectDialog.h (hunk 2 above)
Comment (sk-spr) [reply Userland/DevTools/HackStudio/Dialogs/NewProjectDialog.h:24]: Thank you, I had done that for another thing, I will revert it.

File: Userland/DevTools/HackStudio/HackStudioWidget.cpp (hunk 3 above)
Comment (sk-spr) [reply Userland/DevTools/HackStudio/HackStudioWidget.cpp:1613]: Of course.

File: Userland/DevTools/HackStudio/Dialogs/NewProjectDialog.cpp (hunk 1 above)
Comment (thankyouverycool) [reply Userland/DevTools/HackStudio/Dialogs/NewProjectDialog.cpp:97]: The action can simply warn on unsaved changes before constructing the dialog. This is the standard workflow

File: Userland/DevTools/HackStudio/HackStudioWidget.h (hunk 4 above)
Comment (sk-spr) [reply Userland/DevTools/HackStudio/HackStudioWidget.h:85]: Yes, with the above changes it can be. With my suboptimal solution this needed to be public.

**Comment by sk-spr** [comment]
I have just discovered another problem, I will fix it tomorrow evening.

**Review by ADKaster — COMMENTED** [review]
These seem probably fine. No general C++ sins. I'm not as familiar on the modification state of the editors though.

**Pull request merged**